### PR TITLE
Update LR scheduler configuration 

### PR DESCRIPTION
**Pull request by xiyang-aads-lilly**
This PR is based on https://github.com/microsoft/DeepSpeed/issues/5726.

The current lr scheduler initialization always prioritize config over manual defined scheduler in the code. However, the optimizer initialization implementation prioritize manual defined optimizer over config. This PR aims to make initialization behavior for both optimizer and lr scheduler consistent where if lr scheduler is defined in the code, then it will overwrite config.

### CODE REVIEW

**Comment by xiyang-aads-lilly** [comment]
@microsoft-github-policy-service agree

File: tests/unit/runtime/test_ds_initialize.py [hunk 1]
```diff
@@ -305,3 +305,50 @@ def _lr_scheduler_callable(optimizer) -> _LRScheduler:
...
+            client_scheduler = LambdaLR(client_optimizer, _my_lambda)
+        else:
+            client_scheduler = _lr_scheduler_callable
+
+        _, _, _, ds_lr_scheduler = deepspeed.initialize(config=config_dict,
+                                                        model=model,
+                                                        model_parameters=list(model.parameters()),
+                                                        optimizer=client_optimizer,
+                                                        lr_scheduler=client_scheduler)
+        if client_scheduler is None:
```
Comment (tjruwase) [inline tests/unit/runtime/test_ds_initialize.py:349]: I think a more rigorous test would be to check that each of the three conditions of `scheduler_type` in line 337-342 is correctly handled by the `deepspeed.initialize()` call

```suggestion
        if scheduler_type is None:
```

Comment (xiyang-aads-lilly) [reply tests/unit/runtime/test_ds_initialize.py]: done with the new commit.

**Comment by xiyang-aads-lilly** [comment]
add line 339 to fix test failure

File: tests/unit/runtime/test_ds_initialize.py [hunk 2]
```diff
@@ -305,3 +305,53 @@ def _lr_scheduler_callable(optimizer) -> _LRScheduler:
...
+        hidden_dim = 10
+        model = SimpleModel(hidden_dim)
+
+        client_optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+
+        if scheduler_type is None:
+            config_dict['scheduler'] = {'type': WARMUP_LR, 'params': {}}
+            client_scheduler = None
+        elif scheduler_type == _LRScheduler:
+            client_scheduler = LambdaLR(client_optimizer, _my_lambda)
```
Comment (tjruwase) [inline tests/unit/runtime/test_ds_initialize.py:341]: Thanks for this PR. It looks good to me. But I was wondering if we could improve coverage by using different lr scheduler for `_LRScheduler` and `Callable` options. For example, we could use a different StepLR for `_LRScheduler` option like below. The post-initialization validations should be modified appropriately. What do you think?

```suggestion
            client_scheduler = torch.optim.lr_scheduler.StepLR(client_optimizer)
```

Comment (xiyang-aads-lilly) [reply tests/unit/runtime/test_ds_initialize.py:341]: sure.  I will update and commit again

Comment (xiyang-aads-lilly) [reply tests/unit/runtime/test_ds_initialize.py:341]: Below is what I proposed for the updated testing.  I added two extra functions to test on stepLR and one cycle LR.
Ant suggestions?

```python
@pytest.mark.parametrize("scheduler_type", [None, _LRScheduler, Callable])
class TestClientLrSchedulerInit(DistributedTest):
    world_size = 1

    def test_with_LambdaLR(self, scheduler_type):
        """
        Expect behavior

        if lr scheduler is defined in code and passed into initialize as arg,
        it will be used even this is a lr scheduler has been defined in config.

        Initialize lr scheduler from config when no lr scheduler is defined in code.
        """

        def _my_lambda(epoch):
            return epoch // 10

        def _lr_scheduler_callable(optimizer) -> _LRScheduler:
            return LambdaLR(optimizer, _my_lambda)

        config_dict = {'train_batch_size': 1}

        hidden_dim = 10
        model = SimpleModel(hidden_dim)

        client_optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

        if scheduler_type is None:
            config_dict['scheduler'] = {'type': WARMUP_LR, 'params': {}}
            client_scheduler = None
        elif scheduler_type == _LRScheduler:
            client_scheduler = LambdaLR(client_optimizer, _my_lambda)
        else:
            client_scheduler = _lr_scheduler_callable

        _, _, _, ds_lr_scheduler = deepspeed.initialize(config=config_dict,
                                                        model=model,
                                                        model_parameters=list(model.parameters()),
                                                        optimizer=client_optimizer,
                                                        lr_scheduler=client_scheduler)
        if scheduler_type is None:
            # in this case, we initialize from config
            assert not isinstance(ds_lr_scheduler, LambdaLR)
            assert isinstance(ds_lr_scheduler, WarmupLR)
        else:
            # in this case, we initialize from passed-in scheduler
            assert isinstance(ds_lr_scheduler, LambdaLR)
            assert not isinstance(ds_lr_scheduler, WarmupLR)

    def test_with_StepLR(self, scheduler_type):

        from torch.optim.lr_scheduler import StepLR

        def _lr_scheduler_callable(optimizer) -> _LRScheduler:
            return StepLR(optimizer, step_size=30)

        config_dict = {'train_batch_size': 1}

        hidden_dim = 10
        model = SimpleModel(hidden_dim)

        client_optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

        if scheduler_type is None:
            config_dict['scheduler'] = {'type': WARMUP_LR, 'params': {}}
            client_scheduler = None
        elif scheduler_type == _LRScheduler:
            client_scheduler = StepLR(optimizer, step_size=30)
        else:
            client_scheduler = _lr_scheduler_callable

        _, _, _, ds_lr_scheduler = deepspeed.initialize(config=config_dict,
                                                        model=model,
                                                        model_parameters=list(model.parameters()),
                                                        optimizer=client_optimizer,
                                                        lr_scheduler=client_scheduler)
        if scheduler_type is None:
            assert not isinstance(ds_lr_scheduler, StepLR)
            assert isinstance(ds_lr_scheduler, WarmupLR)
        else:
            assert isinstance(ds_lr_scheduler, StepLR)
            assert not isinstance(ds_lr_scheduler, WarmupLR)

    def test_with_OneCycleLR(self, scheduler_type):

        from deepspeed.runtime.lr_schedules import OneCycle, ONE_CYCLE, CYCLE_MIN_LR, CYCLE_MAX_LR
        from torch.optim.lr_scheduler import OneCycleLR

        def _lr_scheduler_callable(optimizer) -> _LRScheduler:
            return OneCycleLR(optimizer, max_lr=0.01, total_steps=200)

        config_dict = {'train_batch_size': 1}

        hidden_dim = 10
        model = SimpleModel(hidden_dim)

        client_optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

        if scheduler_type is None:
            config_dict['scheduler'] = {'type': ONE_CYCLE, 'params': {CYCLE_MIN_LR: 0, CYCLE_MAX_LR: 0.1}}
            client_scheduler = None
        elif scheduler_type == _LRScheduler:
            client_scheduler = OneCycleLR(optimizer, max_lr=0.01, total_steps=200)
        else:
            client_scheduler = _lr_scheduler_callable

        _, _, _, ds_lr_scheduler = deepspeed.initialize(config=config_dict,
                                                        model=model,
                                                        model_parameters=list(model.parameters()),
                                                        optimizer=client_optimizer,
                                                        lr_scheduler=client_scheduler)
        if scheduler_type is None:
            assert not isinstance(ds_lr_scheduler, OneCycleLR)
            assert isinstance(ds_lr_scheduler, OneCycle)
        else:
            assert isinstance(ds_lr_scheduler, OneCycleLR)
            assert not isinstance(ds_lr_scheduler, OneCycle)
```

Comment (tjruwase) [reply tests/unit/runtime/test_ds_initialize.py:341]: Thanks for extending coverage to StepLR and OneCycleLR. However, I think there is still the problem that `scheduler_type = _LRScheduler` and `scheduler_type = Callable` instantiate the same scheduler.

Comment (xiyang-aads-lilly) [reply tests/unit/runtime/test_ds_initialize.py:341]: I copy this part from `class TestClientLrScheduler(DistributedTest)`(https://github.com/microsoft/DeepSpeed/blob/ffe0af23575c4f03a07408eacfc50b1a58781429/tests/unit/runtime/test_ds_initialize.py#L280-L287) 

If I understand right, what you propose to test is we test different scheduler for `scheduler_type`

for example:

```python
def _lr_scheduler_callable(optimizer) -> _LRScheduler:
            return OneCycleLR(optimizer, max_lr=0.01, total_steps=200)

 if scheduler_type is None:
    config_dict['scheduler'] = {'type': WARMUP_LR, 'params': {}}
    client_scheduler = None
elif scheduler_type == _LRScheduler:
    client_scheduler = LambdaLR(client_optimizer, _my_lambda)
else:
    client_scheduler = _lr_scheduler_callable
```

in this case, we can test the post-initialization validations handles different options.

Just want to get confirmation if I understand right. If so, I will update the code accordingly and re-submit a commit.

Comment (tjruwase) [reply tests/unit/runtime/test_ds_initialize.py:341]: Yes, that is exactly correct. Thanks!

**Review by loadams — APPROVED** [review]
